pkg/controller/jobframework/reconciler.go

```diff
@@ -554,61 +532,14 @@ func generatePodsReadyCondition(job GenericJob, wl *kueue.Workload) metav1.Condi
 	}
 }
 
-func cloneNodeSelector(src map[string]string) map[string]string {
-	ret := make(map[string]string, len(src))
-	for k, v := range src {
-		ret[k] = v
-	}
-	return ret
-}
-
-// getPodSetsInfoFromObjectAnnotation tries to retrieve a podSetsInfo slice from the
-// object's annotations fails if it's not found or is unable to unmarshal
-func getPodSetsInfoFromObjectAnnotation(obj client.Object, job GenericJob) ([]PodSetInfo, error) {
-	hasCounts := true
-	str, found := obj.GetAnnotations()[controllerconsts.OriginalPodSetsInfoAnnotation]
-	if !found {
-		hasCounts = false
-		str, found = obj.GetAnnotations()[controllerconsts.OriginalNodeSelectorsAnnotation]
-		if !found {
-			return nil, errPodSetsInfoNotFound
-		}
-	}
-	// unmarshal
-	ret := []PodSetInfo{}
-	if err := json.Unmarshal([]byte(str), &ret); err != nil {
-		return nil, err
-	}
-
-	if !hasCounts {
-		podSets := job.PodSets()
-		psMap := slices.ToRefMap(podSets, func(ps *kueue.PodSet) string { return ps.Name })
-		for i := range ret {
-			info := &ret[i]
-			ps, found := psMap[info.Name]
-			if !found {
-				return nil, fmt.Errorf("%w: %s", errUnknownPodSetName, info.Name)
-			}
-			info.Count = ps.Count
+// getPodSetsInfoFromWorkload retrieve the podSetsInfo slice from the
+// provided workload's spec
+func getPodSetsInfoFromWorkload(wl *kueue.Workload) []PodSetInfo {
+	return slices.Map(wl.Spec.PodSets, func(ps *kueue.PodSet) PodSetInfo {
```

Comment:
While this is technically possible, go is not a functional language.

I prefer we use a for loop for consistency.

Reply:
it's just passing a function pointer, which happens to be anonymous. The pattern is maybe not be used a lot because is not useful most of the times, but in this case it is.

Reply:
It saves a couple of lines of code, but it's not that easy to read.
I'm on the fence.